AP_Motors: move Thrust Linearization functions and variables to own class

IamPete1 opened the pull request:
Moves functions and code to new class for easier use by others in the future. https://github.com/ArduPilot/ardupilot/pull/22661

Since I was moving the code I also took the opportunity for motors ask baro for air density ratio directly rather than having the vehicles set it into motors.

bnsgeyer commented:
@IamPete1 I have pulled this branch into a branch on my repo.   I was trying to add the class and declare the thr_lin object.  I keep getting a error: redefinition of 'Class Thrust_Linearization'.  This happens with just trying to put the include file and the declaration
Thrust_Linearization thr_lin {*this};  

Tried this in AP_MotorsHeli_Single.h and then I tried it in AP_MotorsHeli.h.  I get the same error.  Not sure how to fix this.  Would appreciate you help.  Thanks!

IamPete1 commented:
@bnsgeyer Opps. I forgot `#pragma once` at the top of `AP_Motors_Thrust_Linearization.h`. I have added it and rebased.

bnsgeyer commented:
@IamPete1 Great!  thanks.  That did the trick and the new implementation made putting it into heli very easy.  Do you think we need to take this back to the DevCall for review or to plug testing

rmackay9 commented:
In general this looks good to me.

IamPete1 commented:
I have pulled the motor example changes in from https://github.com/ArduPilot/ardupilot/pull/22854. This allowed to run a mixer sweep and compare PWM outputs. The outputs remains unchanged. Mostly, there are some 1949's that become 1950's this is due to the removal of the `f` on constants ef `1.0f` -> `1.0`, so the maths comes out slightly more accurate. 

So the expo/min/max bit is the same. But the battery and altitude scaling is not tested in that test. 

I'm not really sure how else to test. Obviously vehicles seems to still fly fine in the sim but that is not the most nuanced.

rmackay9 commented:
This was discussed at the dev call and we decided to ask @IamPete1 to do a bit more testing to ensure the airpressure and voltage changes affect the thrust.

IamPete1 commented:
Rebased.

I can't think of a nice way to allow us to fake up battery and baro readings in the example for a automated sweep. I'm just going to hard code values and compare manually.

IamPete1 commented:
Manual compare complete, no change. I ran three baro and three battery points as well as repeated the test without either.
```
baro_0-5 Max diff: 1 Mean: 0.0101869
baro_1-0 Max diff: 1 Mean: 0.00722728
baro_1-25 Max diff: 1 Mean: 0.00117551
bat_3-3 Max diff: 0 Mean: 0
bat_3-75 Max diff: 1 Mean: 0.00194363
bat_4-2 Max diff: 1 Mean: 0.00722728
stock Max diff: 1 Mean: 0.00722728
```
Data files and compare script here:
[ThrustLinComp.zip](https://github.com/ArduPilot/ardupilot/files/11237699/ThrustLinComp.zip)

lthall commented:
I am happy for this to go in given it performs identically as before. However, I am not able to talk with any authority on how it was done :)

Thanks for all your effort cleaning up the motors library Pete!!!

bnsgeyer commented:
@IamPete1 thanks for all of the work you did on this!!  It will be great to get this into heli for DDFP tailrotors.

IamPete1 commented on libraries/AP_Motors/examples/AP_Motors_test/wscript:
```diff
@@ -2,6 +2,9 @@
 
 def build(bld):
+
+    bld.env.DEFINES.append("AP_MOTORS_DENSITY_COMP=0")
```
This is currently failing CI because this define is being used in the main build, being here I would have expected it to only be used in the example.

IamPete1 replied:
Moved to using build type in the cpp

IamPete1 commented:
Removed the define from the example `wscript` that was polluting everything, moved to using build type to disable density comp in the example. (baro is not there in the example)
